akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -1138,25 +1164,41 @@ class DDataShardCoordinator(
 
     case UpdateTimeout(AllShardsKey, Some(newShard: String)) =>
       log.error(
-        "The ShardCoordinator was unable to update shards distributed state within 'updating-state-timeout': {} millis (retrying), event={}",
+        "The ShardCoordinator was unable to update shards distributed state within 'updating-state-timeout': {} millis ({}), event={}",
         writeMajority.timeout.toMillis,
+        if (terminating) "terminating" else "retrying",
         evt)
-      // repeat until UpdateSuccess
-      sendAllShardsUpdate(newShard)
+      if (terminating) {
+        context.stop(self)
+      } else {
+        // repeat until UpdateSuccess
+        sendAllShardsUpdate(newShard)
+      }
 
     case ModifyFailure(key, error, cause, _) =>
       log.error(
         cause,
-        "The ShardCoordinator was unable to update a distributed state {} with error {} and event {}.Coordinator will be restarted",
+        "The ShardCoordinator was unable to update a distributed state {} with error {} and event {}. {}",
         key,
         error,
-        evt)
-      throw cause
+        evt,
+        if (terminating) "Coordinator will be terminated due to Terminate message received"
+        else "Coordinator will be restarted")
+      if (terminating) {
+        context.stop(self)
+      } else {
+        throw cause
+      }
 
     case GetShardHome(shard) =>
       if (!handleGetShardHome(shard))
         stash() // must wait for update that is in progress
 
+    case ShardCoordinator.Internal.Terminate =>
+      log.debug("Received termination message while waiting for update")
+      terminating = true
+      stash()
```

Comment:
I find the fact that we both toggle and stash here a bit trickier than it needs to be, 

Given that we have the flag we have saved the fact that we got it, so when modify completes (successfully or with a failure) or times out we know we are stopping and can do that, so skip stash here and do it in the same way for all three cases instead?

Reply:
The idea was to continue processing more messages already in the stash when the update completes with UpdateSuccess, but stop if it completes with UpdateFailure.

Reply:
Ah, I see. Thanks for clarifying.